lib/puppet/agent.rb

```diff
@@ -1,6 +1,167 @@
 require 'puppet/application'
 require 'puppet/error'
 
+# A class providing pre/post fork callbacks modeled after those registered
+# by the `pthread_atfork` function.
+# The callbacks defined in this class are no-ops on all platforms except
+# Solaris where they ensure the forked process runs in a different contract
+# than the parent process. This is necessary in order for the child process
+# (an agent run) to be able to restart/stop the puppet agent service without
+# being killed (as a result of the termination of the parent process
+# contract during the service restart/stop).
+# The implementation uses Fiddle to make calls to functions defined in the
+# libcontract library.
+class AtFork
+  require 'fiddle'
+
+  raise LoadError, %q(The loaded fiddle version doesn't define 'Fiddle::Handle') unless defined?(Fiddle::Handle)
+
+  private
+
+  {
+    'libcontract.so.1' => [
+      #function name,            return value type, parameter types, ...
+      [:ct_ctl_abandon,          Fiddle::TYPE_INT,  Fiddle::TYPE_INT],
+      [:ct_tmpl_activate,        Fiddle::TYPE_INT,  Fiddle::TYPE_INT],
+      [:ct_tmpl_clear,           Fiddle::TYPE_INT,  Fiddle::TYPE_INT],
+
+      [:ct_tmpl_set_informative, Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+      [:ct_tmpl_set_critical,    Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+      [:ct_pr_tmpl_set_param,    Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+      [:ct_pr_tmpl_set_fatal,    Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+
+      [:ct_status_read,          Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT, Fiddle::TYPE_VOIDP],
+
+      [:ct_status_get_id,        Fiddle::TYPE_INT,  Fiddle::TYPE_VOIDP],
+      [:ct_status_free,          Fiddle::TYPE_VOID, Fiddle::TYPE_VOIDP],
+    ],
+  }.each do |library, functions|
+    begin
+      libhandle = Fiddle::Handle.new(library)
+    rescue Fiddle::DLError => e
+      raise LoadError, e.message, e.backtrace
+    end
+
+    functions.each do |f|
+      define_method f[0], Fiddle::Function.new(libhandle[f[0].to_s], f[2..-1], f[1]).method(:call).to_proc
+    end
+  end
+
+  CTFS_PR_ROOT = File.join('', %w(system contract process))
+  CTFS_PR_TEMPLATE = File.join(CTFS_PR_ROOT, %q(template))
+  CTFS_PR_LATEST = File.join(CTFS_PR_ROOT, %q(latest))
+
+  CT_PR_PGRPONLY = 0x4
+  CT_PR_EV_HWERR = 0x20
+
+  CTD_COMMON = 0
+
+  def activate_new_contract_template()
+    begin
+      tmpl = File.open(CTFS_PR_TEMPLATE, File::RDWR)
+    rescue
+      Puppet.fatal 'Failed to open new contract template'
+    end
+
+    begin
+      tmpl_fd = tmpl.fileno
+
+      raise if ct_pr_tmpl_set_param(tmpl_fd, CT_PR_PGRPONLY) != 0
+      raise if ct_pr_tmpl_set_fatal(tmpl_fd, CT_PR_EV_HWERR) != 0
+      raise if ct_tmpl_set_critical(tmpl_fd, 0) != 0
+      raise if ct_tmpl_set_informative(tmpl_fd, CT_PR_EV_HWERR) != 0
+
+      raise if ct_tmpl_activate(tmpl_fd) != 0
+    rescue
+      tmpl.close()
```

Comment:
I don't think you can avoid that problem (the code above has the same problem if someone adds code between the line with the `File.open` call and the `begin` on the the next line) unless at the expense of more clutter:
```
tmpl = nil
begin
  tmpl = File.open(...)
  ...
rescue
  tmpl.close() unless tmpl.nil?
end
```